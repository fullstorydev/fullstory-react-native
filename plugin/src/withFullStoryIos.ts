import {
  withDangerousMod,
  withPlugins,
  withXcodeProject,
  withInfoPlist,
  ConfigPlugin,
} from "@expo/config-plugins";
import { mergeContents } from "@expo/config-plugins/build/utils/generateCode";

import { FullStoryIosProps } from ".";

const fs = require("fs");
const path = require("path");

async function readFileAsync(path: string) {
  return fs.promises.readFile(path, "utf8");
}

async function saveFileAsync(path: string, content: string) {
  return fs.promises.writeFile(path, content, "utf8");
}

const withInfoPlistDelegate: ConfigPlugin<FullStoryIosProps> = (
  config,
  { org, host, recordOnStart }
) =>
  withInfoPlist(config, (config) => {
    config.modResults.FullStory = {
      OrgId: org,
      Host: host,
      RecordOnStart: recordOnStart,
    };
    return config;
  });

const withBuildPhaseDelegate: ConfigPlugin = (config) =>
  withXcodeProject(config, (config) => {
    const xcodeProject = config.modResults;

    const fullStoryBuildPhase = xcodeProject.pbxItemByComment(
      "Run FullStory Asset Uploader",
      "PBXShellScriptBuildPhase"
    );

    if (!fullStoryBuildPhase) {
      xcodeProject.addBuildPhase(
        [],
        "PBXShellScriptBuildPhase",
        "Run FullStory Asset Uploader",
        null,
        {
          shellPath: "/bin/sh",
          shellScript:
            "${PODS_ROOT}/FullStory/tools/FullStoryCommandLine " +
            "${CONFIGURATION_BUILD_DIR}/${WRAPPER_NAME}",
        }
      );
    }

    return config;
  });

export function addFullStoryToPodfile(src: string, version: string) {
  return mergeContents({
    tag: "@fullstory/react-native podfile",
    src,
    newSrc: `pod 'FullStory', :http => 'https://ios-releases.fullstory.com/fullstory-${version}-xcframework.tar.gz'`,
    anchor: /use_expo_modules!/,
    offset: 0,
    comment: "#",
  }).contents;
}

const withPodfileDelegate: ConfigPlugin<FullStoryIosProps> = (
  config,
  { version }
) =>
  withDangerousMod(config, [
    "ios",
    async (config) => {
      const file = path.join(config.modRequest.platformProjectRoot, "Podfile");
      const contents = await readFileAsync(file);
      await saveFileAsync(file, addFullStoryToPodfile(contents, version));
      return config;
    },
  ]);

const withFullStoryIos: ConfigPlugin<FullStoryIosProps> = (
  config,
  pluginConfigs
) => {
  return withPlugins(config, [
    [withInfoPlistDelegate, pluginConfigs],
    [withPodfileDelegate, pluginConfigs],
    withBuildPhaseDelegate,
  ]);
};

export default withFullStoryIos;
