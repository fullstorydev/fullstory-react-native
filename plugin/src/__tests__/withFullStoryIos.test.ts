import fs from "fs/promises";
import path from "path";
import { addFullStoryToPodfile } from "../withFullStoryIos";

const pluginConfigs = require("./fixtures/fullstoryConfig.json");

describe("Config Plugin iOS Tests", function () {
  let podfile: string;

  beforeAll(async function () {
    podfile = await fs.readFile(path.resolve(__dirname, "./fixtures/Podfile"), {
      encoding: "utf-8",
    });
  });

  it("Adds FullStory to Podfile", async function () {
    let result = podfile;
    result = addFullStoryToPodfile(result, pluginConfigs.version);
    expect(result).toMatchSnapshot();
  });
});
