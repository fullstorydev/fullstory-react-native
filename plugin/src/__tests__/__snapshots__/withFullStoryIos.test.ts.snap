// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Plugin iOS Tests Adds FullStory to Podfile 1`] = `
"require File.join(File.dirname(\`node --print \\"require.resolve('expo/package.json')\\"\`), \\"scripts/autolinking\\")
require File.join(File.dirname(\`node --print \\"require.resolve('react-native/package.json')\\"\`), \\"scripts/react_native_pods\\")
require File.join(File.dirname(\`node --print \\"require.resolve('@react-native-community/cli-platform-ios/package.json')\\"\`), \\"native_modules\\")

platform :ios, '12.0'

require 'json'
podfile_properties = JSON.parse(File.read('./Podfile.properties.json')) rescue {}

target 'HelloWorld' do
# @generated begin @fullstory/react-native podfile - expo prebuild (DO NOT MODIFY) sync-e0c5a00239acd8e9abee7fd89110259318af5457
pod 'FullStory', :http => 'https://ios-releases.fullstory.com/fullstory-1.27.1-xcframework.tar.gz'
# @generated end @fullstory/react-native podfile
  use_expo_modules!
  config = use_native_modules!

  use_react_native!(
    :path => config[:reactNativePath],
    :hermes_enabled => podfile_properties['expo.jsEngine'] == 'hermes'
  )

  # Uncomment to opt-in to using Flipper
  #
  # if !ENV['CI']
  #   use_flipper!('Flipper' => '0.75.1', 'Flipper-Folly' => '2.5.3', 'Flipper-RSocket' => '1.3.1')
  # end

  post_install do |installer|
    react_native_post_install(installer)

    # Workaround \`Cycle inside FBReactNativeSpec\` error for react-native 0.64
    # Reference: https://github.com/software-mansion/react-native-screens/issues/842#issuecomment-812543933
    installer.pods_project.targets.each do |target|
      if (target.name&.eql?('FBReactNativeSpec'))
        target.build_phases.each do |build_phase|
          if (build_phase.respond_to?(:name) && build_phase.name.eql?('[CP-User] Generate Specs'))
            target.build_phases.move(build_phase, 0)
          end
        end
      end
    end
  end

  post_integrate do |installer|
    begin
      expo_patch_react_imports!(installer)
    rescue => e
      Pod::UI.warn e
    end
  end

end
Footer
"
`;
